Data race in module resolver — SIGSEGV in `loadModuleFromTargetExportOrImport` / `OrderedMap.Keys` under concurrent goroutines

## Stack trace

```
unexpected fault address 0x7263736c65616883
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x2 addr=0x7263736c65616883 pc=0x1046e62c4]

goroutine 125 gp=0x769df0ffe780 m=42 mp=0x769df1601008 [running]:
runtime.throw({0x104c351d4?, 0x769df0d20b20?})
	runtime/panic.go:1229 +0x38
runtime.sigpanic()
	runtime/signal_unix.go:945 +0x21c
github.com/microsoft/typescript-go/internal/module.(*resolutionState).loadModuleFromTargetExportOrImport(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:823 +0x1f64
github.com/microsoft/typescript-go/internal/module.(*resolutionState).loadModuleFromTargetExportOrImport-range1(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:838
github.com/microsoft/typescript-go/internal/module.(*resolutionState).loadModuleFromTargetExportOrImport.(*OrderedMap[...]).Keys.func2(...)
	github.com/microsoft/typescript-go/internal/collections/ordered_map.go:138
... (recursive frames through loadModuleFromTargetExportOrImport) ...
github.com/microsoft/typescript-go/internal/module.(*resolutionState).loadModuleFromExports(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:682 +0x2e4
github.com/microsoft/typescript-go/internal/module.(*resolutionState).loadModuleFromSpecificNodeModulesDirectory(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:1129 +0x678
github.com/microsoft/typescript-go/internal/module.(*resolutionState).loadModuleFromImmediateNodeModulesDirectory(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:1028 +0xa8
github.com/microsoft/typescript-go/internal/module.(*resolutionState).loadModuleFromNearestNodeModulesDirectoryWorker.func1(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:1009 +0x8c
github.com/microsoft/typescript-go/internal/tspath.ForEachAncestorDirectory[...](...)
	github.com/microsoft/typescript-go/internal/tspath/path.go:1068 +0x4c
github.com/microsoft/typescript-go/internal/module.(*resolutionState).loadModuleFromNearestNodeModulesDirectoryWorker(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:1004 +0x40
github.com/microsoft/typescript-go/internal/module.(*resolutionState).loadModuleFromNearestNodeModulesDirectory(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:989 +0x1bc
github.com/microsoft/typescript-go/internal/module.(*resolutionState).resolveNodeLikeWorker(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:563 +0x3b4
github.com/microsoft/typescript-go/internal/module.(*resolutionState).resolveNodeLike(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:514 +0x94
github.com/microsoft/typescript-go/internal/module.(*Resolver).ResolveModuleName(...)
	github.com/microsoft/typescript-go/internal/module/resolver.go:304 +0x750
github.com/microsoft/typescript-go/internal/compiler.(*fileLoader).resolveImportsAndModuleAugmentations(...)
	github.com/microsoft/typescript-go/internal/compiler/fileloader.go:554 +0x6fc
github.com/microsoft/typescript-go/internal/compiler.(*parseTask).load(...)
	github.com/microsoft/typescript-go/internal/compiler/filesparser.go:158 +0x624
github.com/microsoft/typescript-go/internal/compiler.(*filesParser).start.func1()
	github.com/microsoft/typescript-go/internal/compiler/filesparser.go:290 +0x308
github.com/microsoft/typescript-go/internal/core.(*parallelWorkGroup).Queue.func1()
	github.com/microsoft/typescript-go/internal/core/workgroup.go:40 +0x24
sync.(*WaitGroup).Go.func1()
	sync/waitgroup.go:258 +0x48
runtime.goexit()
	runtime/asm_arm64.s:1447 +0x4
created by sync.(*WaitGroup).Go in goroutine 1
```

## Steps to reproduce

```bash
# pnpm monorepo of ~62 workspaces, each with a package.json script:
#   "code-check:tsgo": "tsgo -p ./tsconfig.json --tsBuildInfoFile ./.tsbuildinfo.tsgo"

# Top-level recursive runner (default --workspace-concurrency=4):
pnpm -r --no-bail --aggregate-output code-check:tsgo

# 0–4 random workspaces fail per run. Re-running picks a different set.
# Single-workspace runs in isolation are 100% reliable.
```

NOTE: This was working perfectly fine for us until not long ago. I don't know which version introduced it, but its broken with `7.0.0-dev.20260518.1` and the latest `7.0.0-dev.20260527.1`

## Environment

- macOS 26.5
- Apple M4 Max, 16 cores
- Node v24.15.0
- pnpm 11.3.0

## Workaround

running with `GOMAXPROCS=1` seems to make it work, albeit 3x slower.

~UPDATE: `GOMAXPROCS=2` seems to work fine as well, and retains most of the speed.~
UPDATE 2: Still have issues with `GOMAXPROCS=2`

I can reproduce what looks like the same issue.
## Environment

- macOS arm64
- Node `v24.16.0`
- pnpm `11.5.0`
- `@typescript/native-preview` `7.0.0-dev.20260527.2`

## Repro Shape

The command being looped is a direct `tsgo` check for one package:

```bash
for i in {1..100}; do
  echo "--- run $i ---"
  tsgo --noEmit -p path/to/package/tsconfig.json 2>&1 | tee "./tsgo-run/tsgo-run-$i.log"
  test "${PIPESTATUS[0]}" -eq 0 || true
done
```

The result is nondeterministic. Across 30 runs with a freshly downloaded upstream `@typescript/native-preview-darwin-arm64@7.0.0-dev.20260527.2` binary I saw:

- 22 clean passes
- 5 incorrect TypeScript diagnostic failures
- 3 SIGSEGV crashes

I also saw the same class of failures when this package was checked as part of a larger build graph, but the direct single-package loop above is enough to reproduce both the incorrect diagnostics and the SIGSEGVs.

## Diagnostic failure mode

The non-crashing failures look like random module resolution loss, usually for `effect` packages that are definitely installed and that pass in adjacent runs:

```text
Cannot find module 'effect/Config' or its corresponding type declarations.
Cannot find module 'effect/Tracer' or its corresponding type declarations.
Cannot find module 'effect' or its corresponding type declarations.
Cannot find module 'effect/Layer' or its corresponding type declarations.
Cannot find module 'effect/unstable/http/Headers' or its corresponding type declarations.
Cannot find module 'effect/unstable/http/HttpServerRequest' or its corresponding type declarations.
Cannot find module 'effect/unstable/http/HttpClientResponse' or its corresponding type declarations.
Cannot find module '@effect/vitest' or its corresponding type declarations.
```

## Crash failure modes

The crashes happened while checking the same package, but not on every run.

One crash was in parser code:

```text
unexpected fault address 0x6f632d69612f78a1
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x2 addr=0x6f632d69612f78a1 pc=0x100eba70c]

github.com/microsoft/typescript-go/internal/parser.(*Parser).parseExpectedMatchingBrackets
github.com/microsoft/typescript-go/internal/parser.(*Parser).parseBlock
github.com/microsoft/typescript-go/internal/parser.(*Parser).parseFunctionBlock
```

The stack dump includes what looks like path bytes in memory:

```text
/Users/.../node_modules/.pnpm/effect@4.0.0-beta.74/node_modules/effect/dist/Layer.d.ts
```

Another crash happened while GC was scanning, but active goroutines were in TypeScript-Go file parsing/loading:

```text
SIGSEGV: segmentation violation
PC=0x104ed4288 m=0 sigcode=2 addr=0x2f73656c75646f81

runtime.scanObject
github.com/microsoft/typescript-go/internal/compiler.(*filesParser).parse
github.com/microsoft/typescript-go/internal/compiler.processAllProgramFiles
github.com/microsoft/typescript-go/internal/compiler.NewProgram
github.com/microsoft/typescript-go/internal/execute.performIncrementalCompilation
```

There are also many goroutines around:

```text
github.com/microsoft/typescript-go/internal/compiler.(*fileLoader).resolveImportsAndModuleAugmentations
github.com/microsoft/typescript-go/internal/compiler.(*parseTask).load
github.com/microsoft/typescript-go/internal/compiler.(*filesParser).start.func1
github.com/microsoft/typescript-go/internal/core.(*parallelWorkGroup).Queue.func1
```

## Notes

This looks very similar to this issue because:

- The failure is nondeterministic.
- The crash and incorrect diagnostics both point at concurrent file loading/module resolution/parsing.
- The fault addresses look like corrupted path/string data, e.g. chunks resembling `node_modules`.

I can provide the full logs privately if useful.

--- a/internal/vfs/osvfs/realpath_darwin.go
+++ b/internal/vfs/osvfs/realpath_darwin.go
@@ -20,9 +20,8 @@ import (
 //   - fcntl(fd, F_GETPATH, buf) asks the kernel for the canonical path of the
 //     open file descriptor, written into a MAXPATHLEN buffer.
 //
-// unix.FcntlInt takes an int arg, but on darwin amd64/arm64 Go's int is
-// 64 bits — the same width as a pointer — so the buffer address
-// round-trips through int without loss.
+// unix.FcntlInt takes an int arg, so call it through a uintptr-escaping wrapper
+// to keep the buffer pointer valid until fcntl returns.
 
 var hasFGetPath = sync.OnceValue(func() bool {
 	// Verify that F_GETPATH is supported by this kernel version.
@@ -38,8 +37,13 @@ var hasFGetPath = sync.OnceValue(func() bool {
 
 func fcntlGetPath(fd int, buf *[unix.PathMax]byte) (int, error) {
 	return ignoringEINTR(func() (int, error) {
-		return unix.FcntlInt(uintptr(fd), unix.F_GETPATH, int(uintptr(unsafe.Pointer(&buf[0]))))
+		return fcntlGetPathPtr(uintptr(fd), uintptr(unsafe.Pointer(&buf[0])))
 	})
+}
+
+//go:uintptrescapes
+func fcntlGetPathPtr(fd uintptr, buf uintptr) (int, error) {
+	return unix.FcntlInt(fd, unix.F_GETPATH, int(buf))
 }
 
 func realpath(path string) (string, error) {

